pkg/serviceaccounttoken/secret.go

```diff
@@ -164,3 +188,57 @@ func isSecretForServiceAccount(secret *v1.Secret, sa *v1.ServiceAccount) bool {
 	annotation := annotations[serviceAccountSecretAnnotation]
 	return sa.Name == annotation
 }
+
+func acquireLease(ctx context.Context, clientSet kubernetes.Interface, namespace, name string) error {
+	leaseClient := clientSet.CoordinationV1().Leases(namespace)
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprint(LeasePrefix + name),
+			Namespace: namespace,
+		},
+		Spec: coordinationv1.LeaseSpec{
+			HolderIdentity:       stringPtr("serviceaccounttoken-controller"),
+			LeaseDurationSeconds: int32Ptr(30),
+		},
+	}
+	// Wait for the Lease to be granted
+	backoff := wait.Backoff{
+		Duration: 1 * time.Second,
+		Cap:      100 * time.Millisecond,
+		Steps:    50,
+	}
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		_, err := leaseClient.Create(ctx, lease, metav1.CreateOptions{})
+		// if create was success, meaning we got the lease
+		if err == nil {
+			return true, nil
+		}
+		// if lease already exists, another request has this lease, continue to wait
+		if errors.IsAlreadyExists(err) {
+			return false, nil
+		}
+		return false, err
+	})
+	if err != nil {
+		return fmt.Errorf("error acquiring the lease for %v: %w", name, err)
+	}
+	return nil
+}
+
+func releaseLease(ctx context.Context, clientSet kubernetes.Interface, namespace, name string) error {
+	leaseClient := clientSet.CoordinationV1().Leases(namespace)
+	err := leaseClient.Delete(ctx, fmt.Sprint(LeasePrefix+name), metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("error deleting lease: %w", err)
+	}
+	return nil
+}
+
+// Helper functions for getting pointers
+func stringPtr(s string) *string {
```

Comment:
You could replace these with a generic function:

```go
func valPtr[T any](v T) *T {
        return &v
}
```

Can we add tests for this change?

Reply:
UPDATE: I forgot about https://pkg.go.dev/k8s.io/utils/ptr

You can use `ptr.To()` from there.

Reply:
Fixed